### ISSUE TITLE
Prettify.js is now abandoned/archived. highlight.js as a replacment?

### ISSUE THREAD

**Issue by WizardGed** [PrivateBin/PrivateBin#835]
As per title the google code-prettify library that is used by this project has been archived and is no longer maintained. (see https://github.com/googlearchive/code-prettify). I briefly looked at what other projects are starting to do and most seem to be moving to [highlight.js](https://github.com/highlightjs/highlight.js/), I know swapping out libraries especially libraries that affect visual styling can be controversial so I thought I'd reach out before taking a shot at swapping out the library and making a pull. Is this an acceptable replacement?, Is there a specific way you want the libraries added (clone the cdn library as a git submodule under vendors with the update script, then copy the js libraries to the relevant CSS and JS directories? just copy the min.js file and css files?) I took a look through the faq and didn't find anything regarding this and don't want to make a mess of a pull.

**Comment by elrido** [comment]
A replacement for the prettify library would be welcome and with the goal of the next release being UI change focused, we can anyway expect a few breaking changes (i.e. the upgrade of bootstrap 3 to 5 will change the default UI quite significantly).

Technically: We just put the minified (if these are offered) JS and CSS libriaries with a version identifier in the filename (so the browser cache automatically gets refreshed when a new version is used) into the relevant directories. Some users use the git repository straight on their servers and our main release method is `git archive`, so we try to avoid committing unnecessary bits from the libraries project (see for example the [.gitignore rules to avoid committing composer artefacts](https://github.com/PrivateBin/PrivateBin/blob/cb5ba022bac86279849be74ee5a7b7253b14487f/.gitignore#L15-L27) that aren't used by us).

UI considerations: In my personal opinion, I'd wish for the replacement to also have some kind of support for dark themes (be that a dark-mode or via [themes that need to be configured, like prettify](https://github.com/PrivateBin/PrivateBin/wiki/Configuration#syntaxhighlightingtheme)) and more importantly, that it works without having to tell it what language the code is. If the source code type style (optionally) can be hinted, i.e. from a drop-down or such, the paste format would need to be extended to store this information (presumably via a new field at the end of the [paste_meta_json](https://github.com/PrivateBin/PrivateBin/wiki/Encryption-format#data-passed-in) list, so that older client versions can just ignore it). Pastes created before this change would not have that information and should still be displayed in some form, even if it is just line numbers or it defaults to a generic language like C or POSIX shell scripts.

Related topics, that may or may not influence this replacement:
- #142 wished for being able to disable syntax highlighting, while keeping line numbers. If a highlight.js implementation adds selectable syntaxes, maybe it also includes a "plain text" mode?
- #189 asked for a syntax highlight mode for diff/patch files. Maybe highlight.js already supports this.
- #789 works on letting paste users highlight, lines or sections in a paste. Presumably this requires the presence of line numbers.

**Comment by rugk** [comment]
First of all, @WizardGed, thanks a lot for your enthusiasm to jump in and wrap out legacy/abandoned dependencies! IMHO it is a good idea to upgrade to a proper library.

In addition to what @elrido already said, …
* https://github.com/PrivateBin/PrivateBin/issues/225 is about what @elrido explained with switching the . https://highlightjs.org/ was also suggested there.
* and we could fix https://github.com/PrivateBin/PrivateBin/issues/421 by doing that

**Comment by Tigerpanzer02** [comment]
Hey, what about using [prism.js](https://prismjs.com/) instead of [highlight.js](https://highlightjs.org/) ? 
Prsim got a lot of plugins available which would allow new features and different ways of how it is handled now. 

#142 Possible by not adding x plugin as user
#189 https://prismjs.com/plugins/diff-highlight/
#789 https://prismjs.com/plugins/line-highlight/ & https://prismjs.com/plugins/line-numbers/
#225 is possible 
#421 is fixed with prsimjs [example on prsimjs](https://prismjs.com/test.html#language=python&text=import%20os%0A%0Ahelpstring%20%3D%20'''sss%20bla%20%0Aokay%20newline%0Ahkjhjas%20don't%20end%20it%20here''')

**Comment by elrido** [comment]
Go for it!

**Comment by rugk** [comment]
FYI in the future (with good browser support) we or libraries we use could switch to the _Custom Highlight API_: https://codepen.io/bramus/full/VwRqGVo https://css-tricks.com/css-custom-highlight-api-early-look/#aa-enter-the-css-custom-highlight-api

**Comment by cryptomilk** [comment]
The best syntax highlighting you get in the meantime from tree-sitter: https://tree-sitter.github.io/tree-sitter/syntax-highlighting

It is what Github is using for syntax highlighting.. For the web see: https://github.com/tree-sitter/tree-sitter/blob/master/lib/binding_web/README.md

**Comment by elrido** [comment]
A problem with a lot of these alternatives is that they don't do language auto-detection or rather do operate differently depending on selected language. This means we would almost have to extend the paste format to include a setting for the language in source code type pastes and add the languages to the format drop down (and in the background split the source code ones into two pieces of information). Existing pastes would then probably not get rendered correctly, though that is a minor issue.

At least tree-sitter does offer a regex based mechanism to configure detection patterns, but I've not found samples of how to use that or curated lists of detection patterns. So it certainly has a higher chance of developing a drop-in replacement, though it will still be a significant piece of work.

**Comment by cryptomilk** [comment]
I think this is fine. My cpaste command line client does source code language detection. So I can send the guessed language to the privatebin. Also you can just allow to change the hightlighting. Not only set it when creating the paste, allow to change it when displaying it. Then in case of the wrong highlight someone can change it on display.

**Comment by rugk** [comment]
Yeah, language detection could also be it's own library though and we just pass in the result of that detection into the 
Or (/And), indeed, we could provide a Dropdown and a possibility to choose the language manually. (Especially as some auto detection errors are likely unavoidable – for short content, at least.)

**Comment by elrido** [comment]
I'm all for user choice and this would also elegantly solve the backwards compatibility problem reg. pastes that don't include a (to be defined, future) source language attribute. Good call!

**Comment by BlythT** [comment]
> Hey, what about using [prism.js](https://prismjs.com/) instead of [highlight.js](https://highlightjs.org/) ? Prsim got a lot of plugins available which would allow new features and different ways of how it is handled now.
> 
> [#142](https://github.com/PrivateBin/PrivateBin/issues/142) Possible by not adding x plugin as user [#189](https://github.com/PrivateBin/PrivateBin/issues/189) https://prismjs.com/plugins/diff-highlight/ [#789](https://github.com/PrivateBin/PrivateBin/pull/789) https://prismjs.com/plugins/line-highlight/ & https://prismjs.com/plugins/line-numbers/ [#225](https://github.com/PrivateBin/PrivateBin/issues/225) is possible [#421](https://github.com/PrivateBin/PrivateBin/issues/421) is fixed with prsimjs [example on prsimjs](https://prismjs.com/test.html#language=python&text=import%20os%0A%0Ahelpstring%20%3D%20'''sss%20bla%20%0Aokay%20newline%0Ahkjhjas%20don't%20end%20it%20here''')

I'm working on a branch to replace Prettify.js with prism.js :)

Still refining it, but it knocks out a bunch of these issues in one go: https://github.com/PrivateBin/PrivateBin/pull/1882

I will update the PR body and post here with screenshots to show off what I've implemented.

**Comment by cryptomilk** [comment]
Why not use tree-sitter instead?

**Comment by BlythT** [comment]
Fair point, I'll give tree-sitter a go and see how it compares on benchmarks and visuals

**Comment by cryptomilk** [comment]
We can ship parsers for the most common languages like C, Rust, Python, HTML, CSS, JSON, XML, diff/patch. And we could allow administrators then to add more if the want.
